src/linker/Linker.Dataflow/MethodBodyScanner.cs

```diff
@@ -258,47 +293,53 @@ public virtual void InterproceduralScan (MethodBody methodBody)
-				//Debug.Assert (methodsInGroup.Count () == 1 + calleeMethods.Count ());
-				//foreach (var method in calleeMethods)
-				//	Debug.Assert (methodsInGroup.Contains (method));
+				// Debug.Assert (interproceduralState.Count == 1 + calleeMethods.Count ());
+				// foreach (var method in calleeMethods)
+				// 	Debug.Assert (interproceduralState.Any (kvp => kvp.Key.Method == method));
 			} else {
-				Debug.Assert (methodsInGroup.Count () == 1);
+				Debug.Assert (interproceduralState.Count == 1);
 			}
 #endif
-
-			bool TryGetNextMethodToScan ([NotNullWhen (true)] out MethodDefinition? method)
-			{
-				foreach (var candidate in methodsInGroup) {
-					var candidateMethod = candidate.Method;
-					if (!scannedMethods.Contains (candidateMethod) && candidateMethod.HasBody) {
-						method = candidateMethod;
-						return true;
-					}
-				}
-				method = null;
-				return false;
-			}
 		}
 
-		void TrackNestedFunctionReference (MethodReference referencedMethod, ref ValueSet<MethodProxy> methodsInGroup)
+		void TrackNestedFunctionReference (MethodReference referencedMethod, ref HoistedLocalStore hoistedLocals, ref InterproceduralState interproceduralState)
 		{
 			if (_context.TryResolve (referencedMethod) is not MethodDefinition method)
 				return;
 
 			if (!CompilerGeneratedNames.IsLambdaOrLocalFunction (method.Name))
 				return;
 
-			methodsInGroup = MethodLattice.Meet (methodsInGroup, new (method));
+			// On entry to a nested function, save the current state of hoisted locals (which might be accessed
+			// by the nested function). Note that the input HoistedLocalStore stores ValueBasicBlockPairs for the
+			// current function, whereas the interprocedural state tracks HoistedLocalState which doesn't include basic
+			// block numbers. It only tracks the possible values on entry to the nested function, regardless of the
+			// basic block that assigned the values.
+			var nestedFunctionHoistedLocalState = new HoistedLocalState (defaultValue: UnknownValue.Instance);
+			foreach (var (hoistedLocal, valueBasicBlock) in hoistedLocals)
+				nestedFunctionHoistedLocalState.Set (hoistedLocal, valueBasicBlock.Value);
+
+			interproceduralState.Set (new MethodProxy (method),
+				InterproceduralStateLattice.ValueLattice.Meet (
+					interproceduralState.Get (new MethodProxy (method)),
+					new Maybe<HoistedLocalState> (nestedFunctionHoistedLocalState)));
 		}
 
-		protected virtual void Scan (MethodBody methodBody, ref ValueSet<MethodProxy> methodsInGroup)
+		protected virtual void Scan (MethodBody methodBody, ref InterproceduralState interproceduralState)
 		{
 			MethodDefinition thisMethod = methodBody.Method;
 
 			LocalVariableStore locals = new (methodBody.Variables.Count);
+			// Initialize the hoisted locals state on entry to this method, from the interprocedural state.
+			// Use -1 as the basic block index for any values we track on entry to the method.
+			var defaultValue = new ValueBasicBlockPair (UnknownValue.Instance, -1);
```

Comment:
This is only used right below - can you please inline it? (It's confusing - "default value" can have many meanings)